File: algorytmy/sortowanie/q_sort.c
/**
 * Przyjmuje adresy zmiennych
 */
void int_swap(int * a, int * b)
{
    int buff = *a;
    (*a) = (*b);
    (*b) = buff;
}

// quick_sort
/// ===========================================
int partition(int * arr, int * p, int * q)
{
    // LOG_TAB(p, q - p + 1, "FRAG:");

    // na  pivot wybieramy ostatni element
    int pvt = *q;
    // ustawiamy się adres wcześniej
    int * i = p - 1; 

    for (int * j = p; j < q; ++j)
        if ((*j) <= pvt)
            int_swap(++i, j);
    
    int_swap(++i, q);

    // LOG_TAB(p, q - p + 1, "RES:");

    return i - arr;
}

/**
 * Sortuje przekazaną tablicę (lub jej wyszczególniony fragment)
 * \param arr - wsk, do pierwszego elementu tablicy do posortowania
 * \param p - początek przedziału do posortowania (indeks)
 * \param q - koniec przedziału do posortowania (indeks)
 */
void q_sort(int * arr, int p, int q)
{
    if (p < q)
    {
        int pvt = partition(arr, arr + p, arr + q);
        q_sort(arr, p, pvt - 1);
        q_sort(arr, pvt + 1, q);
    }
}
/// ===========================================

File: zadania/ASD/01_26.02.2020/qsort.py
from random import random


def partition(arr, p, q):
    pivot = arr[q]
    i = p - 1

    for j in range(p, q):
        if arr[j] <= pivot:
            i += 1
            arr[j], arr[i] = arr[i], arr[j]

    arr[i + 1], arr[q] = arr[q], arr[i + 1]
    return i + 1


def qsort(arr, p, q):
    if p < q:
        x = partition(arr, p, q)
        qsort(arr, p, x - 1)
        qsort(arr, x + 1, q)


if __name__ == "__main__":
    arr = [int(random() * 1000) for i in range(0, 10)]
    print(arr)
    qsort(arr, 0, len(arr) - 1)
    print(arr)
